### ISSUE TITLE
Wrong Selector TypeScript definition?

### ISSUE THREAD

**Issue by theCuriousOne** [reactjs/reselect#322]
I believe this definition

     export type Selector<S, R> = (state: S) => R;

 should be changed to:

export type Selector<S, P, R> = (state: S, props: P) => R;

since according to the docs you support the props https://github.com/reactjs/reselect#accessing-react-props-in-selectors

**Comment by aikoven** [comment]
What you're looking for is `ParametricSelector` type, see [src/index.d.ts#L11](https://github.com/reactjs/reselect/blob/v3.0.1/src/index.d.ts#L11).

**Comment by radix** [comment]
I recently updated typescript and I'm getting errors related to this. I think some behavior may have changed.

For some reason my selector functions are no longer type checking, with TS telling me that the props passed to my selector might be undefined. It seems to be unioning these two possibilities (Selector and ParametricSelector), maybe? I'm using createSelectorCreator. I haven't yet figured out a workaround. 

```
    TS2345: Argument of type 'OutputParametricSelector<PTUI, { scene: Scene; }, { creatures: Creature[]; combat: Combat | undefined; }, (res1: App, res2: Scene) => { creatures: Creature[]; combat: Combat | undefined; }>' is not assignable to parameter of type 'MapStateToPropsParam<{ creatures: Creature[]; combat: Combat | undefined; }, { scene: Scene; }>'.
  Type 'OutputParametricSelector<PTUI, { scene: Scene; }, { creatures: Creature[]; combat: Combat | undefined; }, (res1: App, res2: Scene) => { creatures: Creature[]; combat: Combat | undefined; }>' is not assignable to type 'MapStateToPropsFactory<{ creatures: Creature[]; combat: Combat | undefined; }, { scene: Scene; }>'.
    Types of parameters 'props' and 'ownProps' are incompatible.
      Type '{ scene: Scene; } | undefined' is not assignable to type '{ scene: Scene; }'.
        Type 'undefined' is not assignable to type '{ scene: Scene; }'.
```

this is happening with a selector of `(_: any, props: { scene: T.Scene }) => props.scene`. I'm not sure why, all of a sudden, TS is telling me that `undefined` might be passed as props.

**Comment by aikoven** [comment]
@radix Could you post a snippet of code where it fails?

**Comment by radix** [comment]
Ok, I'll try to make a minimal reproducer.

**Comment by radix** [comment]
Of course, after working on a minimized version I learned my assumptions were incorrect :)

I don't think the problem with props being optional was coming from reselect. I think this may be an issue with react-redux, but I haven't confirmed that yet.
